src/Router/RouteCollectorDelegator.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio\Router;
+
+use Laminas\ServiceManager\Factory\DelegatorFactoryInterface;
+use Mezzio\MiddlewareFactoryInterface;
+use Psr\Container\ContainerInterface;
+use Webmozart\Assert\Assert;
+
+use function assert;
+use function is_a;
+use function is_array;
+use function is_string;
+
+final class RouteCollectorDelegator implements DelegatorFactoryInterface
+{
+    /**
+     * RouteCollector Delegator
+     *
+     * Delegates around the RouteCollectorInterface and triggers all registered route providers prior to returning the
+     * RouteCollector instance.
+     *
+     * @inheritDoc
+     */
+    public function __invoke(
+        ContainerInterface $container,
+        $name,
+        callable $callback,
+        ?array $options = null
+    ): RouteCollectorInterface {
+        $collector = $callback();
+        Assert::isInstanceOf($collector, RouteCollectorInterface::class);
+
+        $config = $container->get('config');
+        Assert::isArray($config);
+
+        $routerConfig = $config['router'] ?? [];
+        assert(is_array($routerConfig));
+        $providers = $routerConfig['route-providers'] ?? [];
+        assert(is_array($providers));
+
+        $middlewareFactory = $container->get(MiddlewareFactoryInterface::class);
+        foreach ($providers as $provider) {
+            assert(is_string($provider));
+            assert(is_a($provider, RouteProviderInterface::class, true));
+
+            $instance = $container->has($provider)
+                ? $container->get($provider)
+                : new $provider();
```

Comment:
I think we shouldn't `new $provider()`, ever: it's a shortcut, but declaring services is the way, IMO.